mig-agent/api/api.go

```diff
@@ -0,0 +1,148 @@
+package api
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+// Each endpoint format string assumes that the base address for the API will
+// be something like "https://domain.com/api/v1".
+const (
+	heartbeatEndptFmt string = "%s/heartbeat"
+)
+
+// Client is an HTTP client abstraction for the MIG API.
+// It must be configured with:
+//
+// 1. The base address of the API, such as https://api.mig.mozilla.com/api/v1
+// 2. A list of addresses of proxies to try to use.
+//
+// Note that the base address MUST NOT contain a trailing forward-slash (/).
+// Note also that the client will try each proxy in order.
+type Client struct {
+	baseAddress    string
+	proxyAddresses []string
+}
+
+// Heartbeat contains information describing an active agent.
+type Heartbeat struct {
+	Name        string      `json:"name"`
+	Mode        string      `json:"mode"`
+	Version     string      `json:"version"`
+	PID         uint        `json:"pid"`
+	QueueLoc    string      `json:"queueLoc"`
+	StartTime   time.Time   `json:"startTime"`
+	RefreshTime time.Time   `json:"refreshTime"`
+	Environment Environment `json:"environment"`
+	Tags        []Tag       `json:"tags"`
+}
+
+// Tag is a label associated with an agent.
+type Tag struct {
+	Name  string `json:"name"`
+	Value string `json:"value"`
+}
+
+// Environment contains information about the environment an agent is running in.
+type Environment struct {
+	Init      string   `json:"init"`
+	Ident     string   `json:"ident"`
+	OS        string   `json:"os"`
+	Arch      string   `json:"arch"`
+	IsProxied bool     `json:"isProxied"`
+	Proxy     string   `json:"proxy"`
+	Addresses []string `json:"addresses"`
+	PublicIP  string   `json:"publicIP"`
+	Modules   []string `json:"modules"`
+}
+
+// NewClient constructs a client that can be used to make requests to the
+// MIG API through proxies.
+func NewClient(baseAddr string, proxies []string) Client {
+	return Client{
+		baseAddress:    baseAddr,
+		proxyAddresses: proxies,
+	}
+}
+
+// Heartbeat posts a heartbeat message to the MIG API, indicating that the
+// agent is active.
+func (client Client) Heartbeat(hb Heartbeat) error {
+	fullURL := fmt.Sprintf(heartbeatEndptFmt, client.baseAddress)
+	encoded, _ := json.Marshal(hb)
+	body := bytes.NewReader(encoded)
+	response := struct {
```

Comment:
In this case, I believe response will only get loaded with the error message, not the Response Body from the HTTP POST. Is that intended? Would changing 'response' to 'responseErr' clarify that? Since sendJSON is generalized, should we allow for returning the full response (status, error, body, etc.)

Reply:
`sendJSON` is actually JSON-deserializing the body of the response it gets into whatever container you pass to it. So in this case, since `response` is a struct with a field tagged for JSON (de)serialization, it will expect to deserialize the body as a JSON object containing an `error` field of type `string | null`.